source/fab/steps/compile_fortran.py

```diff
@@ -70,111 +91,212 @@ def run(self, artefact_store, config):
 
-            # results are not the same instances as passed in, due to mp copying
-            compiled_fpaths = {i.analysed_file.fpath for i in compiled_this_pass}
-            all_compiled.extend(compiled_this_pass)
-            already_compiled_files.update(compiled_fpaths)
+        # what can we compile next?
+        compile_next = self.get_compile_next(compiled, uncompiled)
 
-            # remove from remaining to compile
-            to_compile = set(filter(lambda af: af.fpath not in compiled_fpaths, to_compile))
+        # compile
+        logger.info(f"\ncompiling {len(compile_next)} of {len(uncompiled)} remaining files")
+        results_this_pass = self.run_mp(items=compile_next, func=self.process_file)
+        check_for_errors(results_this_pass, caller_label=self.name)
+        compiled_this_pass = list(by_type(results_this_pass, CompiledFile))
+        logger.debug(f"compiled {len(compiled_this_pass)} files")
 
-        log_or_dot_finish(logger)
-        logger.debug(f"compiled per pass {per_pass}")
-        logger.info(f"total fortran compiled {sum(per_pass)}")
+        # hash the modules we just created
+        new_mod_hashes = get_mod_hashes(compile_next, config)
+        self._mod_hashes.update(new_mod_hashes)
 
-        if to_compile:
-            logger.debug(f"there were still {len(to_compile)} files left to compile")
-            for af in to_compile:
-                logger.debug(af.fpath)
-            logger.error(f"there were still {len(to_compile)} files left to compile")
-            exit(1)
+        # add compiled files to all compiled files
+        compiled.update({cf.input_fpath: cf for cf in compiled_this_pass})
 
-        # add the targets' new object files to the artefact store
-        lookup = {compiled_file.analysed_file.fpath: compiled_file for compiled_file in all_compiled}
-        target_object_files = artefact_store.setdefault(COMPILED_FILES, defaultdict(set))
-        for root, source_files in build_lists.items():
-            new_objects = [lookup[af.fpath].output_fpath for af in source_files]
-            target_object_files[root].update(new_objects)
+        # remove compiled files from remaining files
+        uncompiled = set(filter(lambda af: af.fpath not in compiled, uncompiled))
+        return uncompiled
 
-    def get_compile_next(self, already_compiled_files: Set[Path], to_compile: List[AnalysedFile]):
+    def get_compile_next(self, compiled: Dict[Path, CompiledFile], uncompiled: Set[AnalysedFile]) -> Set[AnalysedFile]:
 
         # find what to compile next
         compile_next = set()
         not_ready: Dict[Path, List[Path]] = {}
-        for af in to_compile:
+        for af in uncompiled:
             # all deps ready?
-            unfulfilled = [dep for dep in af.file_deps if dep not in already_compiled_files and dep.suffix == '.f90']
+            unfulfilled = [dep for dep in af.file_deps if dep not in compiled and dep.suffix == '.f90']
```

Comment:
It might be me who is missing something. Let's explore!

If a file depends on a module which doesn't exist, then it needs Fab to compile the file which produces that module. Fab detects module dependencies and turns them into file dependencies for the compilation stage, as the only thing that matters is the order in which the files are compiled.

With this incremental build PR, and _a.f90 --> b.mod_, Fab can tell if _module b_ has changed since _file a_ was last compiled.

Not sure if that was helpful!